src/tir/schedule/primitive/blockize_tensorize.cc

```diff
@@ -0,0 +1,638 @@
+  BlockNode* inner_block = inner_block_realize->block.CopyOnWrite();
+  inner_block_realize->iter_values = extractor.inner_bindings;
+  inner_block_realize->predicate = inner_pred;
+  inner_block->iter_vars = extractor.inner_iter_vars;
+  inner_block->init = NullOpt;
+
+  // Step 6: Generate the outer block.
+  BlockRealize outer_realize =
+      GenerateBlockizedOuterBlock(extractor, block, GetRef<BlockRealize>(inner_block_realize),
+                                  collector.inner_loops, outer_pred, &analyzer);
+  // Step 7: Do the actual replacement
+  self->Replace(loop_sref, outer_realize, {{block, GetRef<Block>(inner_block)}});
+
+  // Step 8: Update the cached flags
+  const StmtSRef& outer_block_sref = self->stmt2ref.at(outer_realize->block.get());
+  StmtSRef scope_root = tir::GetScopeRoot(self, outer_block_sref, /*require_stage_pipeline=*/false,
+                                          /*require_subtree_compact_dataflow=*/false);
+  BlockInfo old_block_info = self->GetBlockInfo(scope_root);
+  self->UpdateScopeBlockInfo(tir::GetBlockRealize(self, scope_root));
+  // 'affine_binding' depends on the outer loops and are not changed.
+  self->block_info[scope_root].affine_binding = old_block_info.affine_binding;
+  return outer_block_sref;
+}
+
+/*!
+ * \brief Update the map from the buffers in the description to the implementation of the tensor
+ * intrinsic. \param intrinsic The tensor intrinsic. \param buffer_map The map to be updated.
+ */
+void RemapTensorIntrinBuffers(
+    const TensorIntrin& intrinsic,
+    std::unordered_map<Buffer, Buffer, ObjectPtrHash, ObjectPtrEqual>* buffer_map) {
+  ICHECK_EQ(intrinsic->description->params.size(), intrinsic->implementation->params.size());
+  for (size_t i = 0; i < intrinsic->description->params.size(); ++i) {
+    const auto& lhs_var = intrinsic->description->params[i];
+    const auto& lhs_buffer = intrinsic->description->buffer_map[lhs_var];
+    const auto& rhs_var = intrinsic->implementation->params[i];
+    const auto& rhs_buffer = intrinsic->implementation->buffer_map[rhs_var];
+    (*buffer_map)[rhs_buffer] = lhs_buffer;
+  }
+}
+
+void Tensorize(ScheduleState self, const StmtSRef& loop_sref, const TensorIntrin& intrinsic) {
+  /*!
+   * Check:
+   *   - Check buffer binding, including type, alignment, shape and etc.
+   *   - Check the sub AST is equal to the description function.
+   *
+   * Mutate:
+   *   - Blockize the sub AST (please refer blockize for details)
+   *   - Bind buffers
+   *   - Mutate the implementation of the tensor intrinsic by replacing its buffers with new
+   *     buffers created via match buffer region.
+   *   - Replace the sub tree with the mutated function.
+   */
+  const auto* loop = loop_sref->StmtAs<ForNode>();
+  CHECK(loop) << "Only support tensorize a loop for now";
+
+  const auto* desc_block_realize =
+      Downcast<BlockRealize>(intrinsic->description->body)->block->body.as<BlockRealizeNode>();
```

Comment:
looks like both desc&impl of the tensor intrin is the root block's children, which is another block. Is there any specific reason that we don't use the root block directly?

Reply:
Updated, we are using root block directly. Previously this was because we need to define root iter vars (the parser will create another block as direct child of root block, because root block can't have iter vars)